### DISCUSSION TITLE
Can a Custom addon, manipulate story parameters?

### DISCUSSION THREAD

**Discussion by arcticShadow** [storybookjs/storybook#25021] [Help] [answered]
### Summary

Context:
I've created a very basic feature flagging system, for our specific use case. Its working well by filtering the stories that are displayed in the sidebar using the sidebar filters.

The flagging relies on a list of stories, as identified by either path, name or title, and they are flagged to not display based on an evaluation of the currently deployed environment. (Can go into more details if anyone is curious) 

The very basic version of the addon, is a file with the content:

```typescript

addons.register(ADDON_ID, (api) => {
 api.experimental_setFilter(ADDON_ID, (item) => {
    const flag = getflagForItem(item);

    if (flag && currentEnv <= flag.env) {
      // flag allows the item, so show the item
      return true;
    } else if (flag && currentEnv > flag.env) {
      // flag denies the item, so hide the item
      return false;
    }
}
```

This is working well so far.

Problem:
I have just noticed that the flagged stories are showing up in the autodocs page (inside the `Stories` block) - In retrospect, this is expected behavior. 

What I have been exploring the possibility, is if I can get a list of stories in the addon.register callback, then iterate it, and use api.updateStory to set `parameters.docs.disable=true`

The issue I have come into, is I cannot find out how to get access to the stories in the register method. 

So far I have explored:

* Using the Storybook hooks to get state (failed, as not a component)
* Trying to access the ManagerContext directly using unsupported react methods (failed, because the register function appears to be called before the context is populated. )
* Just calling the api.updateStory with a had coded ID (failed - did not determine why) 
* Looking through the manager code, to see if there are other addon lifecycles I could tap into (Did not find any)

Any ideas or suggesting on how I could mutate story parameters would be appreciated.

**Comment by JReinhold** [comment, marked as answer]
I don't think dynamically changing parameters will suit your needs, because not everything listens for changes on those parameters, so the UI will not update based on your imperative changes.

I'd instead suggest implementing your own docs page that filters stories based on additional flags.
I've created a POC of it here: https://stackblitz.com/edit/github-whkwkn?file=.storybook%2Ffiltered-docs-stories.tsx

It defines a global custom autodocs template ([as documented here](https://storybook.js.org/docs/writing-docs/autodocs#write-a-custom-template)). The custom template is almost the same as the default, with the exception that `<Stories />` is is replaced with a custom `<FilteredStories />` component.

`FilteredStories` is a copy-paste from [the default `Stories`](https://github.com/storybookjs/storybook/blob/next/code/ui/blocks/src/blocks/Stories.tsx), with an added filter function that filters the stories based on some random criteria that you can modify.

Overwriting the default `Stories` is a bit cumbersome, there's an RFC open to simplify it a bit. Feel free to upvote it if it would help you. https://github.com/storybookjs/storybook/discussions/22493

**Reply by arcticShadow** [reply]
Thanks for the detailed response - this is an angle that I had not considered - overwriting the stories block. I've already got a heavily custom docs page so a bit more customisation won't hurt.

I'm curious about your comment "not everything listens for changes to parameters" I can accept that, as to do so would mean reactive programming on change, which adds a whole other layer of complexity. however with add-ons being initialized so early in the bootstrapping sequence it feels like there could be an opportunity to manipulate stories prior to anything else starting to use them.

Anyways - I'll go down the path of using a custom stories block - 🙏 thanks
